algorithmic_efficiency/checkpoint_utils.py

```diff
@@ -0,0 +1,162 @@
+import os
+from typing import Sequence, Tuple
+
+from absl import logging
+from flax import jax_utils
+from flax.training import checkpoints as flax_checkpoints
+import jax
+import torch
+import torch.distributed as dist
+
+from algorithmic_efficiency import spec
+from algorithmic_efficiency.pytorch_utils import pytorch_setup
+
+USE_PYTORCH_DDP, RANK, DEVICE, N_GPUS = pytorch_setup()
+CheckpointReturn = Tuple[spec.OptimizerState,
+                         spec.ParameterContainer,
+                         spec.ModelAuxiliaryState,
+                         dict,
+                         list,
+                         int,
+                         int]
+
+
+def maybe_restore_checkpoint(framework: str,
+                             optimizer_state: spec.OptimizerState,
+                             model_params: spec.ParameterContainer,
+                             model_state: spec.ModelAuxiliaryState,
+                             train_state: dict,
+                             eval_results: list,
+                             global_step: int,
+                             preemption_count: int,
+                             checkpoint_dir: str) -> CheckpointReturn:
+  if framework == 'jax':
+    opt_state, opt_update_fn = optimizer_state
+  else:
+    opt_state, opt_update_fn = optimizer_state, None
+
+  uninitialized_global_step = -1
+  uninitialized_preemption_count = -1
+  checkpoint_state = dict(
+      model_params=model_params,
+      optimizer_state=opt_state,
+      model_state=model_state,
+      train_state=train_state,
+      eval_results=None,
+      global_step=uninitialized_global_step,
+      preemption_count=uninitialized_preemption_count)
+
+  if framework == 'jax':
+    latest_ckpt = flax_checkpoints.restore_checkpoint(
+        checkpoint_dir, target=checkpoint_state)
+    save_path = os.path.join(checkpoint_dir,
+                             'checkpoint_' + str(latest_ckpt['global_step']))
+  else:
+    latest_ckpt = checkpoint_state
+    save_path = os.path.join(checkpoint_dir, 'checkpoint')
+    if os.path.exists(save_path):
+      latest_ckpt = torch.load(save_path, map_location=DEVICE)
```

Comment:
@runame I perform the `map_location` when loading the checkpoint here.

Reply:
Ah perfect, thanks.